src/Qir/QIR-RT-API-Design-Guidelines.md

```diff
@@ -353,103 +353,57 @@ The plan was to replace the exception throw with the [`exit()`](https://www.cplu
+* [Exceptions and Error Handling](https://isocpp.org/wiki/faq/exceptions) As part of Modern C++ FAQ.
+* Google guidelines recommend against exceptions, but [allow them in Windows code, especially STL](https://google.github.io/styleguide/cppguide.html#Windows_Code).
+* [Vishal Chovatiya Guidelines](http://www.vishalchovatiya.com/7-best-practices-for-exception-handling-in-cpp-with-example/), which may be easier to read.
+This section will describe only exceptions with regards to the runtime API.
+
+* Do not handle hardware exceptions and out-of-memory exception
+
+It is not recommended to catch asynchronous harware exceptions, this is sometimes called
+[Structured Exception Handling](https://docs.microsoft.com/en-us/cpp/cpp/structured-exception-handling-c-cpp?view=msvc-160).
+An example of a hardware exception is a segmentation fault. If such exceptional situation arises, the application should
+fail fast and quit. Typically hardware exceptions are handled in a system level code, and our runtime is an application with
+regards to the host OS. NOTE: MSVC compiler accepts the [`/EHs`](https://docs.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model?view=msvc-160) switch to disable hardware exception handing in `catch` blocks.
+
+It is also not recommended to handle out-of-memory exception as there's a very little chance of handling it right.
+For such handling to work correctly there shouldn't be any memory allocation from the point of exception, in the stack
+unwinding process, till the handling of exception and in the handling of exception until the out-of-memory situation
+is resolved. Just let the application fail. For the same reason, using non-throwing version of operator new is not recommended.
+
+* Mark interface functions with noexcept
+
+If you implement an interface function mark it with the [noexcept specifier](https://en.cppreference.com/w/cpp/language/noexcept_spec):
+
+> void interface_func() **noexcept**
+
+If an exception is thrown (and not caught) within the noexcept function, the application will terminate. Termination will
+ensure that exceptions will never cross the API boundaries. An overhead is incurred if a noexcept function
+calls functions that can throw exceptions, so consider marking all functions that shouldn't throw exceptions as noexcept.
+If you are calling functions that can throw exceptions, make sure to handle them properly.
+
+extern "C" functions aren't considered noexcept by default. Consider marking them noexcept, but only when they are
+compiled by the C++ compiler (and not by a pure C compiler). NOTE: MSVC compiler accepts
+[`/EHsc`](https://docs.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model?view=msvc-160)
+flag to treat extern "C" functions as not throwing.
```

Comment:
I'd suggest the following change (does the code example illustrate what you mean?):
> The `extern "C"` functions are not `noexcept` by default. In the exposed headers consider marking them `noexcept` if they are compiled with the C++ compiler (but not marked as `noexcept` if compiled with C compiler):
```c++
#ifdef __cplusplus
#define NOEXCEPT noexcept     // `noexcept` in C++ only.
extern "C"
{
#else
#define NOEXCEPT      // In C this macro is empty.
#endif

.. // `#include`s
void f() NOEXCEPT;    // Exposed function. It is `noexcept` if the header is included to the C++ source file.

#ifdef __cplusplus
} // extern "C"
#endif
```
If the change above is ok then the fragment below probably stops being applicable (and can be excluded):
> NOTE: MSVC compiler accepts[`/EHsc`](https://docs.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model?view=msvc-160)flag to treat extern "C" functions as not throwing.

Reply:
It's applicable but is less relevant. As we already have links to that page I agree it can be excluded.